cli/cmd/summary.go

```diff
@@ -0,0 +1,74 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+type syncSummary struct {
+	CliVersion          string `json:"cli_version"`
+	DestinationErrors   uint64 `json:"destination_errors"`
+	DestinationName     string `json:"destination_name"`
+	DestinationPath     string `json:"destination_path"`
+	DestinationVersion  string `json:"destination_version"`
+	DestinationWarnings uint64 `json:"destination_warnings"`
+	Resources           uint64 `json:"resources"`
+	SourceErrors        uint64 `json:"source_errors"`
+	SourceName          string `json:"source_name"`
+	SourcePath          string `json:"source_path"`
+	SourceVersion       string `json:"source_version"`
+	SourceWarnings      uint64 `json:"source_warnings"`
+	SyncID              string `json:"sync_id"`
+}
+
+func persistSummary(filename string, summaries []syncSummary) error {
+	// if filename is not specified then we don't need to persist the summary and can return
+	if filename == "" {
+		return nil
+	}
+	err := checkFilePath(filename)
+	if err != nil {
+		return fmt.Errorf("failed to validate summary file path: %w", err)
+	}
+	for _, summary := range summaries {
+		dataBytes, err := json.Marshal(summary)
+		if err != nil {
+			return err
+		}
+		dataBytes = append(dataBytes, []byte("\n")...)
+		err = appendToFile(filename, dataBytes)
+		if err != nil {
+			return fmt.Errorf("failed to append summary to file: %w", err)
+		}
+	}
+	return nil
+}
+
+func appendToFile(fileName string, data []byte) error {
+	f, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	if _, err := f.Write(data); err != nil {
+		f.Close()
+		return err
+	}
+	return f.Close()
+}
+
+func checkFilePath(filename string) error {
+	dirPath := filepath.Dir(filename)
+	_, err := os.Stat(dirPath)
+	if err != nil {
+		// if error is that the directory does not exist, create it
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(dirPath, 0755)
+			if err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
```suggestion
	err := os.MkdirAll(dirPath, 0755)
	if err != nil {
		return err
	}
```

Can't we do this instead to simplify? Looks like `os.MkdirAll` handles the checks for us:

```go
func MkdirAll(path string, perm FileMode) error {
	// Fast path: if we can tell whether path is a directory or file, stop with success or error.
	dir, err := Stat(path)
	if err == nil {
		if dir.IsDir() {
			return nil
		}
		return &PathError{Op: "mkdir", Path: path, Err: syscall.ENOTDIR}
	}
```